src/Features/CSharp/Portable/UsePatternMatching/CSharpAsAndNullCheckDiagnosticAnalyzer.Analyzer.cs

```diff
@@ -0,0 +1,351 @@
+                            var whileStatement = (WhileStatementSyntax)current;
+                            return CheckLoop(whileStatement, whileStatement.Statement, defAssignedWhenTrue);
+
+                        case SyntaxKind.IfStatement:
+                            var ifStatement = (IfStatementSyntax)current;
+                            var oppositeStatement = defAssignedWhenTrue
+                                ? ifStatement.Else?.Statement
+                                : ifStatement.Statement;
+
+                            if (oppositeStatement != null)
+                            {
+                                var dataFlow = _semanticModel.AnalyzeDataFlow(oppositeStatement);
+                                if (dataFlow.DataFlowsIn.Contains(_localSymbol))
+                                {
+                                    // Access before assignment is not safe in the opposite branch
+                                    // as the variable is not definitely assgined at this point.
+                                    // For example:
+                                    //
+                                    //    if (o is string x) { }
+                                    //    else { Use(x); }
+                                    //
+                                    return false;
+                                }
+
+                                if (dataFlow.AlwaysAssigned.Contains(_localSymbol))
+                                {
+                                    // If the variable is always assigned here, we don't need to check
+                                    // subsequent statements as it's definitely assigned afterwards.
+                                    // For example:
+                                    //
+                                    //     if (o is string x) { }
+                                    //     else { x = null; }
+                                    //
+                                    return true;
+                                }
+                            }
+
+                            if (!defAssignedWhenTrue &&
+                                !_semanticModel.AnalyzeControlFlow(ifStatement.Statement).EndPointIsReachable)
+                            {
+                                // Access before assignment here is only valid if we have a negative
+                                // pattern-matching in an if-statement with an unreachable endpoint.
+                                // For example:
+                                //
+                                //      if (!(o is string x)) {
+                                //        return;
+                                //      }
+                                //
+                                //      // The 'return' statement above ensures x is definitely assigned here
+                                //      Console.WriteLine(x);
+                                //
```

Comment:
great comments.

Reply:
was there a test for this?

Reply:
Yes, this part was already implemented, so I just ported it in the walker